# Review 1  
## paper  
Identifying cell populations with scRNASeq.Tallulah S. Andrews, Martin Hemberg , Molecular Aspects of Medicine   25 July 2017   
## batch effect   
Normalization methods are available
for correcting for different sequencing depths across cells and removing batch effects ,Sequencing depth can be corrected using counts/transcripts per  
million, or downsampling. Methods developed specifically for scRNASeq include scran ,which has advantages for datasets with many differentially  
expressed genes, and SCnorm which accounts for different effects of sequencing depth of genes with different expression levels. As discussed above, if  
spike-ins were included in the dataset they may be used for normalization strategies which are robust to differentially expressed genes and preserves   
differences due to total RNA content Batch effects can be statistically removed in experiments where each batch contains cells from multiple biological
conditions and each biological condition is spread across multiple batches, i.e. a “balanced” design. Methods for batch correction include RUVs ComBat  
and linear mixed-modelling.  
# Review 2 
## paper 
An Introduction to the Analysis of Single-Cell RNA-Sequencing Data. Aisha A. AlJanahi, Mark Danielsen, and Cynthia E. Dunbar   
Molecular Therapy: Methods & Clinical Development Vol. 10 September 2018
## Normalization   
Data Normalization and Scaling  
When analyzing sequencing data, normalization to eliminate batch effects is crucial if multiple sequencing runs are to be compared with each other.   
An accepted way to normalize the sequencing data is based on comparisons with housekeeping genes.  
the housekeeping gene method can be inaccurate because these genes are not always present in the same amount in different cell populations.To avoid  
making the assumption that a housekeeping gene is present  in an equal amount in all cells, the scaling can be based on all nondifferentially  
expressed genes in all or some of the cells.   
![overiew of this paper](https://github.com/Rickyzhang1990/during_work/blob/master/paper_and_Algorithm/image/overiew.png)  
# Review 3 
## paper   
Integrative single-cell analysis Tim Stuart , Rahul Satija Nature reviews Genetics   
Integrating single-cell data across experiments   
the joint analysis of data sets collected from different single cells poses a key computational challenge for single-cell biology. This challenge  
echoes similar needs for ‘batch-correction’ techniques for bulk data sets.   Newly developed approaches that can first identify shared biological  
states (for example, matched cell types) across data sets can overcome this challenge (Fig. 4a) and have become an area of rapid analytical development
(Seurat v2 R toolkit)The method first applies canonical correlation analysis (CCA) to identify shared sources of variation between the data sets 
A complementary approach, mnnCorrect, accomplishes similar goals through the innovative application of techniques that have previously been applied to  
shape and pattern matching across images
Both CCA and mnnCorrect enable the integration and pooling of scRNA-seq data sets generated by different laboratories and technologies but from the  
same underlying tissue. 
![cca_mnn.png](https://github.com/Rickyzhang1990/during_work/blob/master/paper_and_Algorithm/image/MNN_CCA.png)  
